README.md
<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/suppdata)](https://cran.r-project.org/package=suppdata)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Build Status](https://api.travis-ci.org/ropensci/suppdata.svg)](https://travis-ci.org/ropensci/suppdata)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/willpearse/suppdata?branch=master&svg=true)](https://ci.appveyor.com/project/willpearse/suppdata)
[![DOI](https://joss.theoj.org/papers/10.21105/joss.00721/status.svg)](https://doi.org/10.21105/joss.00721)
[![](https://badges.ropensci.org/195_status.svg)](https://github.com/ropensci/onboarding/issues/195)
[![codecov](https://codecov.io/gh/ropensci/suppdata/branch/master/graph/badge.svg)](https://codecov.io/gh/ropensci/suppdata)
<!-- badges: end -->

# Loading SUPPlementary DATA into R

William D. Pearse, Daniel Nuest, and Scott Chamberlain

## Overview

The aim of this package is to aid downloading data from published
papers. To download the supplementary data from a PLoS paper, for
example, you would simply type:

```{R}
library(suppdata)
suppdata("10.1371/journal.pone.0127900", 1)
```

...and this would download the first supplementary information (SI) from the paper.

This sort of thing is very useful if you're doing meta-analyses, or
just want to make sure that you know where all your data came from and
want a completely reproducible "audit trail" of what you've done.
It uses [`rcrossref`](https://cran.r-project.org/package=rcrossref) to lookup which journal the article is in.

## How to install and load the package
The version on CRAN is the most stable version. You can install and
load it like this:

```{R}
install.packages("suppdata")
library(suppdata)
```

If you want to load the development version, which probably contains
more features but is not always guaranteed to work, load the `master`
branch from this repository like this:

```{R}
library(devtools)
install_github("ropensci/suppdata")
library(suppdata)
```

This package depends on the packages `httr`, `xml2`,
`jsonlite`, and `rcrossref`.

## Supported publishers and repositories

- [bioRxiv](https://www.biorxiv.org/) (`biorxiv`)
- [Copernicus Publications](https://publications.copernicus.org/) (`copernicus`)
- [DRYAD](https://datadryad.org/) (`dryad`)
- [Ecological Society of Ameria - Ecological Archives](http://esapubs.org/archive/) (`esa_archives` and `esa_data_archives`)
- [Europe PMC](https://europepmc.org/) (`epmc`, multiple publishers from life-sciences upported including BMJ Journals, eLife, F1000Research, Wellcome Open Research, Gates Open Research)
- [figshare](https://figshare.com/) (`figshare`)
- [Journal of Statistical Software](https://www.jstatsoft.org/) (`jstatsoft`)
- [MDPI](https://www.mdpi.com/) (`mdpi`)
- [PeerJ](https://peerj.com/) (`peerj`)
- [PLOS | Public Library of Science](https://www.plos.org/) (`plos`)
- [Proceedings of the royal society Biology (RSBP)](https://rspb.royalsocietypublishing.org/) (`proceedings`)
- [Science](https://www.sciencemag.org/) (`science`)
- [Wiley](https://onlinelibrary.wiley.com/) (`wiley`)

See a list of potential sources at [#2](https://github.com/ropensci/suppdata/issues/2) - requests welcome!

## Contributing

[For more details on how to contribute to the package, check out the
guide in `CONTRIBUTING.MD`](CONTRIBUTING.md).

## A more detailed set of motivations for `suppdata`

`suppdata` is an R package to provide easy, reproducible
access to supplemental materials within R. Thus `suppdata` facilitates
open, reproducible research workflows: scientists re-analyzing
published datasets can work with them as easily as if they were stored
on their own computer, and others can track their analysis workflow
painlessly.

For example, imagine you were conducting an analysis of the evolution
of body mass in mammals. Without `suppdata`, such an analysis would
require manually downloading body mass and phylogenetic data from
published manuscripts. This is time-consuming, difficult (if not
impossible) to make truly reproducible without re-distributing the
data, and hard to follow. With `suppdata`, such an analysis is
straightforward, reproducible, and the sources of the data are clear
because their DOIs are embedded within the code:

```{R}
# Load phylogenetics packages
library(ape)
library(caper)
library(phytools)

# Load suppdata
library(suppdata)

# Load two published datasets
tree <- read.nexus(suppdata("10.1111/j.1461-0248.2009.01307.x", 1))[[1]]
traits <- read.delim(suppdata("E090-184", "PanTHERIA_1-0_WR05_Aug2008.txt", "esa_archives"))

# Merge datasets
traits <- with(traits, data.frame(body.mass = log10(X5.1_AdultBodyMass_g), species=gsub(" ","_",MSW05_Binomial)))
c.data <- comparative.data(tree, traits, species)

# Calculate phylogenetic signal
phylosig(c.data$phy, c.data$data$body.mass)
```

## A guided walk through `suppdata`

The aim of `suppdata` is to make it as easy as possible for you to write reproducible analysis scripts that make use of published data. So let's start with that first, simplest case: how to make use of published data in an analysis.

### Learning by example
Below is an example of an analysis run using `suppdata`. Read through it first, and then we'll go through what all the parts mean.

```{R}
# Load phylogenetics packages
library(ape)
library(caper)
library(phytools)

###############################
# LOAD TWO PUBLISHED DATASETS #
#       USING SUPPDATA        #
###############################
library(suppdata)
tree <- read.nexus(suppdata("10.1111/j.1461-0248.2009.01307.x", 1))[[1]]
traits <- read.delim(suppdata("E090-184", "PanTHERIA_1-0_WR05_Aug2008.txt", "esa_archives"))

# Merge datasets
traits <- with(traits, data.frame(body.mass = log10(X5.1_AdultBodyMass_g), species=gsub(" ","_",MSW05_Binomial)))
c.data <- comparative.data(tree, traits, species)

# Calculate phylogenetic signal
phylosig(c.data$phy, c.data$data$body.mass)
```

This short script loads some `R` packages focused on modelling the evolution of species' traits, then it gets to the "good stuff": using `suppdata`. First, we load the `suppdata` package using `library(suppdata)`. The next line uses a function called `read.nexus`, which loads something called a phylogeny (you might be familiar with this if you're a biologist). Normally, this function would take the location of a file on our hard-drive as a single argument, but now we're giving it the output from a call to the `suppdata` function.

`suppdata` is going to the website of the article whose DOI is _10.1111/j.1461-0248.2009.01307.x_ ([it's this paper by Fritz et al.](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1461-0248.2009.01307.x)), and then taking the first (`1`) supplement from that article. It saves that to a temporary location on your hard-drive, and then gives that location to `read.nexus`. _This works with any function that expects a file at a location on your hard-drive_. What particularly neat is that `suppdata` remembers that it has downloaded that file already (see below for more details), such that you only have to download something once per `R` session.

The second call to `suppdata`, which makes use of `read.delim`, shows two of the potential complexities of `suppdata`. First of all, because some journal publishers store their supplementary materials using numbers and others using specific file-names, `suppdata` takes either a number (like in the first example), or a name (like in the second example) depending on the journal publisher you're taking data from. If you look in the help file for `suppdata`, there is a table outlining those options. Sorry, you've just got to read up on it :-( Secondly, if you're an ecologist you might be familiar with the Ecological Society of America's data archives section. While they've moved over to a new way of storing data more recently, if you're hoping to load an older dataset from that journal you need to give the ESA data archive reference and specify that you're downloading from ESA (as in this example). If you're not an ecologist, don't worry about it, as this doesn't apply to you :D

That's it! You now know all you need to in order to use `suppdata`! The rest of the lines of code merge these datasets together, and then calculate something called _phylogenetic signal_ in these datasets. If you're an evolutionary biologist, those lines might be interesting to you. If you're not, then don't worry about them.

### Caching and saving to a specific directory

Sometimes, you will want to use `suppdata` to build a store of files on your hard-drive. If so, you should know that `suppdata` takes three optional arguments: `cache`, `dir`, and `save.name`. If you specify `cache=FALSE`, it will turn off `suppdata`'s caching of files: this will force it to download your data again. This is mostly useful if you somehow make `suppdata` foul itself up (maybe you hit control-c or stop during a download) and so `suppdata` has only half-downloaded a file, and so thinks it's cached something when it hasn't. If you get an error when using `suppdata`, this is a good thing to try setting.

`dir` specifies a directory where `suppdata` should store files, and `save.name` specifies the name that the file should be saved under when saved. This is useful if you want to make a folder on your computer that contains certain files you use a lot: `suppdata` will cache from this folder if you tell it to, and so you can build up a reproducible selection of data to use inbetween `R` sessions.

[![rofooter](https://ropensci.org/public_images/github_footer.png)](https://ropensci.org)


NEWS.md
# suppdata 1.1-9
- Changed FigShare (again!) to match the new online interface

# suppdata 1.1-7
- Changed FigShare download URL to match recent change online

# suppdata 1.1-6
- Hotfixes for CRAN

# suppdata 1.1-5

- Added fall-back for zip downloads on Windows (thanks Alban Sagouis)
- New errors for moved download URLs
- Moved all help file examples back to \dontrun (at CRAN's request)
- Removed ESA data archives; site down at time of required submission

# suppdata 1.1-4

- FigShare website change --> no longer supports named SI

# suppdata 1.1-3

- Changes to available data on EPMC for a demonstration package
  required a change to the code on CRAN to pass its checks

# suppdata 1.1-0

- More thorough and numerous tests
- New journals added
- New co-author - Daniel Nüst!

# suppdata 1.0.0

- Initial release on CRAN


tests/test-all.R
if(require("testthat"))
    test_check("suppdata")


tests/testthat/test-all.r
library(suppdata)
library(testthat)
library(fulltext)

context("suppdata")

test_that("PLOS works", {
  expect_true(file.exists(suppdata("10.1371/journal.pone.0127900", 1)))
  expect_true(file.exists(suppdata("10.1371/journal.pone.0127900", 1, "plos")))
})

test_that("Figshare works", {
  expect_true(file.exists(suppdata("10.6084/m9.figshare.979288", 1)))
  expect_true(file.exists(suppdata("10.6084/m9.figshare.979288", "analysis.R")))
  expect_identical(attr(suppdata("10.6084/m9.figshare.979288", 1),"suffix"),"R")
})

test_that("ESA archives works", {
  expect_true(file.exists(suppdata("E093-059", "myco_db.csv", "esa_archives")))
  expect_identical(attr(suppdata("E093-059", "myco_db.csv", "esa_archives"),
                        "suffix"), "csv")
  })

test_that("ESA data archives works", {
  expect_true(file.exists(suppdata("E092-201", "MCDB_communities.csv",
                                   "esa_data_archives")))
})

test_that("Science works", {
  expect_true(file.exists(suppdata("10.1126/science.1255768",
                                   "Appendix_BanksLeite_etal.txt")))
})

test_that("'Proceedings of the royal society Biology' (RSBP) works", {
  expect_true(file.exists(suppdata("10.1098/rspb.2015.0338", vol = 282,
                                   issue = 1811, 1)))
})

test_that("'bioRxiv works", {
  expect_true(file.exists(suppdata("10.1101/016386", 1)))
})

test_that("'EPMC works", {
  expect_true(file.exists(suppdata("10.1371/journal.pone.0126524",
                                   "pone.0126524.g005.jpg", "epmc")))
})

test_that("DRYAD works", {
  expect_error(file <- suppdata("10.5061/dryad.34m6j", "datafile.csv"), NA)
  expect_error(data <- read.csv(file), NA)
  expect_equal(dim(data), c(145,49))
  expect_error(file <- suppdata("10.5061/dryad.55610", "Data (revised).txt"),
               NA)
  expect_error(data <- read.delim(file), NA)
  expect_equal(dim(data), c(740,25))
})

test_that("Wiley works", {
  expect_true(file.exists(suppdata("10.1111/ele.12437", si = 1)))
  expect_true(file.exists(suppdata("10.1111/ele.12437", si = 2)))
  expect_true(file.exists(suppdata("10.1002/ece3.1679", si = 2)))
  expect_error(suppdata('10.1111/ele.12437', si = 3))
})

test_that("multiple downloads are handled well", {
  expect_true(all(file.exists(suppdata(c("10.1101/016386", "10.1111/ele.12437"),
                                       si = 1))))
  expect_true(all(file.exists(suppdata(c("10.1101/016386", "10.1111/ele.12437"),
                                       si = 2:1))))
  expect_true(all(file.exists(suppdata(c("10.1101/016386", "10.1111/ele.12437"),
                                       si = 1))))
})

test_that("fulltext functions as input are handled well", {
  skip_if_not_installed("fulltext")
  expect_true(file.exists(suppdata(fulltext::ft_search("beyond the edge with edam", limit = 1), 1)))
  expect_error(suppdata(fulltext::ft_search("beyond the edge with edam"), 1))
  expect_error(suppdata(fulltext::ft_get(c("10.1371/journal.pone.0126524","10.1371/journal.pone.0126524")), 1))
  expect_true(file.exists(suppdata(fulltext::ft_get("10.1371/journal.pone.0126524"), 1)))
})

test_that("PeerJ works", {
  expect_true(all(file.exists(suppdata(c("10.7287/peerj.preprints.26561v1",
                                         "10.7717/peerj.3006"),
                                       si = 1))))
  expect_true(all(file.exists(suppdata(c("10.7287/peerj.preprints.26561v1",
                                         "10.7717/peerj.3006"),
                                       si = "supp-2"))))
  expect_equal(attr(suppdata("10.7717/peerj.3006", si = 1), which = "suffix"), "docx")
  expect_equal(attr(suppdata("10.7287/peerj.preprints.26561", si = 1), which = "suffix"), "csv")
})

test_that("Copernicus works", {
  cop_unzipped_files <- suppdata("10.5194/bg-14-1739-2017", si = 1)
  expect_true(dir.exists(cop_unzipped_files))
  expect_length(list.files(cop_unzipped_files), 8)
  cop_zipfile_only <- suppdata("10.5194/bg-14-1739-2017", si = "bg-14-1739-2017-supplement.zip")
  expect_true(file.exists(cop_zipfile_only))
  expect_false(dir.exists(cop_zipfile_only))
  cop_unzip_dir <- suppdata("10.5194/bg-14-1739-2017", si = 1, save.name = "target-directory")
  expect_true(file.exists(file.path(cop_unzip_dir, "bg-14-1739-2017-supplement-title-page.pdf")))
  expect_match(cop_unzip_dir, "/target-directory$")
  
  csv_file <- suppdata("10.5194/bg-14-1739-2017", si = "Table S1 v2 UFK FOR_PUBLICATION.csv")
  expect_true(file.exists(csv_file))
  csv_data <- read.csv(file = csv_file, skip = 3)
  expect_equal(names(csv_data)[2:3], c("YEAR", "NAD83_X"))
  expect_true(file.exists(suppdata("10.5194/bg-14-1739-2017", si = "Table S1 v2 UFK FOR_PUBLICATION.csv", save.name = "data.csv")))
  expect_error(suppdata("10.5194/bg-14-1739-2017", si = 2), "one supplemental archive")
  expect_error(suppdata("10.5194/bg-14-1739-2017", si = "1"), "file not in zipfile")
  
  cop_pdf_file <- suppdata("10.5194/acp-2016-189", si = 1)
  expect_true(file.exists(cop_pdf_file))
  expect_equal(attr(cop_pdf_file, which = "suffix"), "pdf")
})

test_that("suppdata fails well", {
  expect_error(expect_warning(suppdata('nonsense', 1), "404"), "nonsense")
  expect_error(suppdata("10.6084/m9.figshare.979288", si = c(1,2,3)), "length")
  expect_error(suppdata('10.6084/m9.figshare.979288', 20))
  expect_error(suppdata('10.6084/m9.figshare.979288', "does_exist.csv"))
  expect_error(suppdata("10.7717/peerj.3006", si = 99))
  expect_error(suppdata("10.7717/peerj.3006", si = "si_1"))
  expect_error(suppdata("10.5194/bg-15-3625-2018", si = 1), "No supplement found")
  expect_warning(expect_error(suppdata("10.5194/bg-15-3625-xxxx", si = 1), "Cannot find publisher"), "404")
  expect_error(suppdata("10.5194/bg-14-1739-2017", si = "not a file.csv"), "file not in zipfile")
  expect_error(suppdata("10.5061/dryad.34m6j", si = 999), "character SI info")
  expect_error(suppdata("10.1101/016386", si = "999"), "numeric SI info")
  expect_error(suppdata("10.1371/journal.pone.0126524", si = 999, "epmc"), "character SI info")
  expect_error(suppdata("10.1098/rspb.2015.0338", vol = 282, issue = 1811, "99"), "numeric SI info")
  expect_error(suppdata("10.1126/science.1255768", 999), "character SI info")
  expect_error(suppdata("E092-201", 999, "esa_data_archives"), "character SI info")
  expect_error(suppdata("E093-059", 999, "esa_archives"), "character SI info")
  expect_error(suppdata("10.6084/m9.figshare.979288", FALSE), "'si' must be numeric or character")
  expect_error(suppdata:::.suppdata.figshare("10.6084/m9.figshare.979288", FALSE), "numeric or character SI info")
  expect_error(suppdata("10.1371/journal.pone.0127900", "999"), "numeric SI info")
  expect_error(suppdata("10.1111/ele.12437", "999"), "numeric SI info")
  expect_error(suppdata:::.suppdata.plos("10.1111/ele.12437", 1), "Unrecognised PLoS journal")
  expect_error(suppdata(as.character(c()), "1"), "'x' must contain some data")
  expect_error(suppdata("10.1371/journal.pone.0127900", 1, dir = "/wrong.path"), "must exist")
  expect_error(suppdata(fulltext::ft_search("complete ghibberish", limit = 1), 1), "No DOI found in fulltext search")
})

test_that("file name can be set", {
  filename <- "myfile.R"
  saved_file <- suppdata("10.6084/m9.figshare.979288", "analysis.R", save.name = filename)
  expect_true(file.exists(saved_file))
  expect_equal(basename(saved_file), filename)
})

test_that("output dir can be set", {
  testdir <- file.path(tempdir(), "testdir")
  dir.create(testdir)
  saved_file <- suppdata("10.6084/m9.figshare.979288", "analysis.R", dir = testdir)
  expect_true(file.exists(file.path(testdir, "10.6084_m9.figshare.979288_analysis.R")))
  unlink(testdir)
})

test_that("zipfile contents can be listed", {
  expect_output(suppdata(x = "10.1038/nbt.1883", si = "41598_2018_19799_Fig1_HTML.jpg", list = TRUE), "Files in ZIP")
  expect_error(expect_output(suppdata(x = "10.1038/nbt.1883", si = "1", list = TRUE), "41598_2018_19799_MOESM1_ESM.pdf"))
  expect_equal(capture_output(suppdata(x = "10.1038/nbt.1883", si = "41598_2018_19799_Fig1_HTML.jpg")), "")
  expect_equal(capture_output(suppdata(x = "10.1038/nbt.1883", si = "41598_2018_19799_Fig1_HTML.jpg", list = FALSE)), "")
})


tests/testthat/test-esa.R
# library("suppdata")
library("testthat")

context("ESA (data) archives")

skip_if_not_reachable <- function() {
    tryCatch({
        httr::HEAD('http://esapubs.org')
    },
    error = function(e) {
        skip(message = toString(e))
    }
    )
}

test_that("Accessing specific SI from ESA archives works", {
    skip_on_cran()
    skip_if_not_reachable()
    expect_true(file.exists(suppdata("E093-059", "myco_db.csv", "esa_archives")))
})

test_that("Accessing specific SI from ESA archives has correct suffix for file", {
    skip_on_cran()
    skip_if_not_reachable()
    expect_identical(attr(suppdata("E093-059", "myco_db.csv", "esa_archives"),
                          "suffix"), "csv")
})

test_that("ESA archives fail with numeric SI info", {
    skip_on_cran()
    skip_if_not_reachable()
    expect_error(suppdata("E093-059", 999, "esa_archives"), "character SI info")
})

test_that("ESA data archives works", {
    skip_on_cran()
    skip_if_not_reachable()
    expect_true(file.exists(suppdata("E092-201", "MCDB_communities.csv",
                                     "esa_data_archives")))
})

test_that("ESA data archives fail with numeric SI info", {
    skip_on_cran()
    skip_if_not_reachable()
    expect_error(suppdata("E092-201", 999, "esa_data_archives"), "character SI info")
})


R/xeno.R
#' Xeno-Canto meta-data download wrapper
#'
#' Parameters and meanings are described in Xeno-Canto API
#' documentation (http://www.xeno-canto.org/article/153). This is a
#' very light wrapper! You can download the next page of data using
#' the 'page' parameter; I've yet to write a 'continue downloading'
#' wrapper, in part for fear of over-using the API.
#' @param params named vector of parameters to search Xeno-Canto
#' with. Names should be the parameter type, and values should be the
#' values for those searches.
#' @param verbose whether to print what page (of how many) is being
#' downloaded (default: TRUE)
#' @param types categories of meta-data to return about
#' calls. Defaults are intended to be everything everything; for space
#' reasons you might not want everything.
#' @examples
#' xeno.canto.meta(setNames(c("bearded bellbird", "A"), c("species", "q")))
#' @author Will Pearse
#' @importFrom RCurl getURL
#' @importFrom jsonlite fromJSON
#' @export
xeno.canto.meta <- function(params, verbose=TRUE, types=c("id","gen","sp","ssp","en","rec","cnt","loc","lat","lng","type","file","lic","url","q")){
    #Argument handling
    if(!is.character(params) | !"names" %in% names(attributes(setNames(letters, letters))))
        stop("'params' must be a named character vector")
    if(!is.character(types))
        stop("'types' must be a character vector")
    
    #Download raw data
    url <- paste(names(params), params, sep=":", collapse="&")
    url <- URLencode(gsub("species:", "", url, fixed=TRUE))
    url <- paste0("https://www.xeno-canto.org/api/2/recordings?query=", url)

    #Parse and return
    results <- fromJSON(getURL(url))
    if(verbose)
        cat("Downloading page ", results$page, " of ", results$numPages, "\n")
    return(results$recordings)
}

#' Xeno-Canto data download wrapper
#'
#' Parameters and meanings are described in Xeno-Canto API
#' documentation (http://www.xeno-canto.org/article/153). This is a
#' very light wrapper! You can download the next page of data using
#' the 'page' parameter; I've yet to write a 'continue downloading'
#' wrapper, in part for fear of over-using the API.
#' @param ids vector of ids (*not* URLs), probably taken from
#' \code{\link{xeno.canto.meta}}
#' @param save.name a name for the file to download. If \code{NULL}
#' (default) this will be a combination of the DOI and SI number
#' @param dir directory to save file to. If \code{NULL} (default) this
#' will be a temporary directory created for your files
#' @examples
#' xeno.canto.download(xeno.canto.meta(setNames(c("bearded bellbird",
#'     "A"), c("species", "q")))$id[1])
#' xeno.canto.download("247117")
#' xeno.canto.download(247117:247120)
#' @author Will Pearse
#' @importFrom RCurl getURL
#' @importFrom jsonlite fromJSON
#' @export
xeno.canto.download <- function(ids, save.name=NULL, dir=NULL){
    #Internal wrapper
    .dwn <- function(x)
        .download(paste0("https://www.xeno-canto.org/",x), dir, paste(save.name, x, sep="_"))

                  #Argument handling
    if(!is.null(dir)){
        if(!file.exists(dir))
            stop("'dir' must exist unless NULL")
    } else dir <- tempdir()
    if(is.null(save.name))
        save.name <- gsub(.Platform$file.sep, "_", save.name, fixed=TRUE)
    
    #Download and return
    output <- sapply(ids, .dwn)
    return(output)
}


R/utils.R
#' @importFrom rcrossref cr_works
.suppdata.pub <- function(x){
    if(!is.character(x))
        stop("'x' must be a character")

    #Doing the check here saves one internet call
    if(grepl("figshare", x))
        return("figshare")
    if(grepl("dryad", x))
        return("dryad")
    pub <- rcrossref::cr_works(x)$data

    if(length(pub) < 1)
        stop("Cannot find publisher for DOI: ", x)
    if(pub$prefix == "http://id.crossref.org/prefix/10.0000")
        stop("Cannot find publisher for DOI: ", x)
    
    return(.grep.text(pub$member, "[0-9]+"))
}

.suppdata.func <- function(x) {
    #Check by code, return if found
    output <- switch(x, 
                     "340" = .suppdata.plos,
                     "311" = .suppdata.wiley,
                     "221" = .suppdata.science,
                     "175" = .suppdata.proceedings,
                     "246" = .suppdata.biorxiv,
                     "4443" = .suppdata.peerj,
                     "3145" = .suppdata.copernicus,
                     "1968" = .suppdata.mdpi,
                     "7893" = .suppdata.jstatsoft
                     )
    if(!is.null(output))
        return(output)

    #Check by letter code
    output <- switch(x, 
                     "plos" = .suppdata.plos,
                     "wiley" = .suppdata.wiley,
                     "science" = .suppdata.science,
                     "proceedings" = .suppdata.proceedings,
                     "figshare" = .suppdata.figshare,
                     "biorxiv" = .suppdata.biorxiv,
                     "epmc" = .suppdata.epmc,
                     "dryad" = .suppdata.dryad,
                     "peerj" = .suppdata.peerj,
                     "copernicus" = .suppdata.copernicus,
                     "mdpi" = .suppdata.mdpi,
                     "jstatsoft" = .suppdata.jstatsoft
                     )
    #If all else fails, try EPMC
    if(is.null(output))
        output <- .suppdata.epmc
    return(output)
}

# Internal regexp functions
.grep.url <- function(url, regexp, which=1){
    html <- as.character(GET(url))
    return(.grep.text(html, regexp, which))
}
.grep.text <- function(text, regexp, which=1){
    links <- gregexpr(regexp, text)
    if(which > length(links[[1]]))
        stop(
            "SI number '", which, "' greater than number of detected SIs (",
            length(links[[1]]), ")"
        )
    pos <- as.numeric(links[[1]][which])
    return(substr(text, pos, pos+attr(links[[1]], "match.length")[which]-1))
}
.file.suffix <- function(text, max.length=4){
    suffix <- .grep.text(text, "[a-zA-Z]+$")
    if(nchar(suffix) <= max.length & nchar(suffix) > 0)
        return(suffix)
    return(NA)
}

# Internal download function
#' @importFrom utils download.file
.download <- function(url, dir, save.name, cache=TRUE, suffix=NULL, zip=FALSE){
    destination <- file.path(dir, save.name)
    if(is.null(suffix))
        suffix <- .file.suffix(url, 5)
    
    if(cache==TRUE & file.exists(destination)){
        if(!is.na(suffix))
            attr(destination, "suffix") <- suffix
        return(destination)
    }
    # Must download zips *manually* in binary mode, or Windows machines will error
    if(zip){
        result <- download.file(url, destination, quiet=TRUE, mode="wb")
    } else {
        result <- download.file(url, destination, quiet=TRUE)
    }
    if(result != 0)
        stop("Error code", result, " downloading file;", url, "may not exist")
    
    if(!is.na(suffix))
        attr(destination, "suffix") <- suffix
    return(destination)
}

# Internal unzip function
#' @importFrom utils unzip
.unzip <- function(zip, dir, save.name, cache, si, list=FALSE){
    files <- unzip(zip, list=TRUE)
    if(list){
        cat("Files in ZIP:")
        print(files)
    }
    if(!si %in% files$Name)
        stop("Required file not in zipfile ", zip)
    file <- unzip(zip, si)
    file.rename(file, file.path(dir, save.name))
    suffix <- .file.suffix(si, 5)
    output <- file.path(dir, save.name)
    if(!is.na(suffix))
        attr(output, "suffix") <- suffix
    return(output)
}

# Internal URL 'redirect' function
.url.redir <- function(x)
    return(GET(x)$url)

.tmpdir <- function(dir){
    if(!is.na(dir)){
        if(!file.exists(dir))
            stop("'dir' must exist unless NA")
    } else dir <- tempdir()
    return(dir)
}
.save.name <- function(doi, save.name, file){
    if(is.na(save.name)){
        save.name <- paste(doi,file, sep="_")
        save.name <- gsub(.Platform$file.sep, "_", save.name, fixed=TRUE)
    }
    return(save.name)
}
#Expanding arguments when handling ft-classes
.fix.param <- function(x, param, name){
    if(length(x) != length(param)){
        if((length(x) %% length(param)) != 0)
            stop("length of ", name, " (", length(param),
                 ") is incompatible with 'x' (", length(x), ")")
        param <- rep(param, length(x))
    }
    return(param)
}


R/suppdata.R
#' Download supplementary materials from journals
#'
#' Put a call to this function where you would put a file-path - everything 
#' is cached by default, so you don't have to worry about multiple downloads 
#' in the same session.
#' 
#' @param x One of: vector of DOI(s) of article(s) (a
#'     \code{character}) or ESA-specific article code.
#' @param si number of the supplementary information (SI) to be
#'     downloaded (1, 2, 3, etc.), or (for ESA, Science, and
#'     Copernicus journals) the name of the supplement (e.g.,
#'     "S1_data.csv"). Can be a \code{character} or \code{numeric}.
#' @param from Publisher of article (\code{character}). The default
#'     (\code{auto}) uses crossref (\code{\link[rcrossref]{cr_works}})
#'     to detect the journal's publisher. Specifying the journal can
#'     somewhat speed up your download, or be used to force a download
#'     from EPMC (see details). You *must* specify if downloading from
#'     an ESA journal (\code{esa_data_archives},
#'     \code{esa_archives}). You can only use this argument if
#'     \code{x} is a vector of DOI(s). Must be one of: \code{auto}
#'     (i.e., auto-detect journal; default), \code{plos},
#'     \code{wiley}, \code{science}, \code{proceedings},
#'     \code{figshare}, \code{esa_data_archives}, \code{esa_archives},
#'     \code{biorxiv}, \code{epmc}, \code{peerj}, \code{copernicus},
#'     (Data)\code{dryad}), \code{mdpi}, or \code{jstatsoft}.
#' @param save.name a name for the file to download
#'     (\code{character}). If \code{NA} (default) this will be a
#'     combination of the DOI and SI number
#' @param dir directory to save file to (\code{character}). If
#'     \code{NA} (default) this will be a temporary directory created
#'     for your files
#' @param cache if \code{TRUE} (default), the file won't be downloaded
#'     again if it already exists (in a temporary directory creates,
#'     or your chosen \code{dir})
#' @param vol Article volume (Proceedings journals only;
#'     \code{numeric})
#' @param issue Article issue (Proceedings journals only;
#'     \code{numeric})
#' @param list if \code{TRUE}, print all files within a zip-file
#'     downloaded from EPMC (default: FALSE). This is *very* useful if
#'     using EPMC (see details)
#' @param timeout how long to wait for successful download (default 10
#'     seconds)
#' @param zip if \code{TRUE}, force download in binary format in order
#'     to ensure that zipped files will work on Windows (default
#'     FALSE).

#' @author Will Pearse (\email{will.pearse@usu.edu}) and Scott
#'     Chamberlain (\email{myrmecocystus@gmail.com})
#' @note Make sure that the article from which you're attempting to
#'     download supplementary materials *has* supplementary
#'     materials. 404 errors and 'file not found' errors can result
#'     from such cases.
#' @examples
#' # NOTE: The examples below are flagged as 'dontrun' to avoid
#' # running downloads repeatedly on CRAN servers
#' \dontrun{
#' #Put the function wherever you would put a file path
#' crabs <- read.csv(suppdata("10.6084/m9.figshare.979288", 2))
#'
#' epmc.fig <- suppdata("10.1371/journal.pone.0126524",
#'                        "pone.0126524.s002.jpg", "epmc")
#' #...note this 'SI' is not actually an SI, but rather an image from the paper.
#'
#' #View the suffix (file extension) of downloaded files
#' # - note that not all files are uploaded/stored with useful file extensions!
#' attr(epmc.fig, "suffix")
#' 
#' copernicus.csv <- suppdata("10.5194/bg-14-1739-2017",
#'                            "Table S1 v2 UFK FOR_PUBLICATION.csv",
#'                            save.name = "data.csv")
#' #...note this 'SI' is not an SI but the name of a file in the supplementary information archive.
#' }
#' # (examples not run on CRAN to avoid downloading files repeatedly)
#' @template suppdata
#' @export
#' @importFrom stats setNames
suppdata <- function(x, si,
                     from=c("auto","plos","wiley","science","proceedings",
                            "figshare","esa_data_archives","esa_archives",
                            "biorxiv","epmc", "peerj", "copernicus",
                            "jstatsoft"),
                     save.name=NA, dir=NA, cache=TRUE, vol=NA, issue=NA,
                     list=FALSE, timeout=10, zip=FALSE)
    UseMethod("suppdata")
#' @export
#' @rdname suppdata
suppdata.character <- function(x, si,
                               from=c("auto","plos","wiley","science",
                                      "proceedings","figshare",
                                      "esa_data_archives","esa_archives",
                                      "biorxiv","epmc","peerj", "copernicus",
                                      "jstatsoft"),
                               save.name=NA, dir=NA, cache=TRUE,
                               vol=NA, issue=NA, list=FALSE, timeout=10, zip=FALSE){
    #Basic argument handling
    if(length(x) == 0)
        stop("'x' must contain some data!")
    from <- match.arg(from)
    if(!(is.numeric(si) | is.character(si)))
        stop("'si' must be numeric or character")

    #Multiply argument lengths
    from <- .fix.param(x, from, "from")
    si <- .fix.param(x, si, "si")
    save.name <- .fix.param(x, save.name, "save.name")
    dir <- .fix.param(x, dir, "dir")
    vol <- .fix.param(x, vol, "vol")
    issue <- .fix.param(x, issue, "issue")
    cache <- .fix.param(x, cache, "cache")
    list <- .fix.param(x, list, "list")
    timeout <- .fix.param(x, timeout, "timeout")
    
    ############################
    #Recurse if needed
    # - can't use Recall because of potential argument length problems
    if(length(x) > 1)
        return(setNames(unlist(
            mapply(suppdata.character, x=x,si=si,from=from,save.name=save.name,
                   dir=dir,cache=cache,vol=vol,issue=issue,list=list,
                   timeout=timeout, zip=zip))
           ,x))
    ############################
    #...Do work

    #Setup output directory and filename
    if(!is.na(dir)){
        if(!file.exists(dir))
            stop("'dir' must exist unless NA")
    } else dir <- tempdir()
    if(is.na(save.name)){
        save.name <- paste(x,si, sep="_")
        save.name <- gsub(.Platform$file.sep, "_", save.name, fixed=TRUE)
    }

    #Find publisher, download, and return
    if(from == "auto")
        from <- .suppdata.pub(x)
    func <- .suppdata.func(from)
    return(func(x, si, save.name=save.name, dir=dir, cache=cache,
                vol=vol, issue=issue, list=list, timeout=timeout, zip=zip))
}
